CORS error on Origin Reflection misconfiguration #103 (#105)
xsleaks/wiki: content/docs/attacks/cache-probing.md

Often, some of these methods might be considered a bug in the browser (e.g. [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=959789#c9)).

## CORS error on Origin Reflection misconfiguration

> Origin reflection is a behaviour in which a gobally accessible resource is provided with a [Access-Control-Allow-Orign (ACAO)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin) header value that reflects the origin which initialized the request insted of respond with the wildcard `*`. This can be considered as a [CORS misconfiguration](https://web-in-security.blogspot.com/2017/07/cors-misconfigurations-on-large-scale.html) even if some backend frameworks such as Python Flask and Ruby on Rails promote origin reflection as a default behaviour for a globally accessible API. What happens is that, under the assumption of a resource hosted on target.com, when it is visualized from target.com itself then it is provided with `Access-Control-Allow-Origin: target.com` header. What is important is that this information is stored toghether with the resource in browser cache memory. Now, following the same concepts, if a third-party website called attacker.com fetches the same resource two possible ways can be followed:

Author: Ok tommorrow I will add the snippet and the references required. I have also added @rick.titor as contributor since we did this work together.

> - The resource was never seen before: the resource will be fetched and served and stored toghether with `Access-Control-Allow-Origin: attacker.com` header.
> - The resource was already visualized by the victim: it will generate a cache hit, so it will be found in browser cache memory but it will also generate a CORS error due to the ACAO header value mismatch with the requesting origin. 
> This method provides a highly reliable way to make cache probing since it exploits error events which do not suffer from network performances. The best way to mitigate this is to avoid origin reflection and use the wildcard * for globally accessible resources.
> 
> 

Reviewer: ```suggestion
```

Often, some of these methods might be considered a bug in the browser (e.g. [this bug](https://bugs.chromium.org/p/chromium/issues/detail?id=959789#c9)).

## CORS error on Origin Reflection misconfiguration

Origin reflection is a behavior in which a globally accessible resource is provided with a [Access-Control-Allow-Orign (ACAO)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin) header whose value reflects the origin that initialized the request. This can be considered as CORS misconfiguration [^5] and can be used to detect whether the resource exists in the browser cache.

{{< hint info >}} 

> As an example Python Flask promote origin reflection as the default behavior for a globally accessible API.[^6] 

Reviewer: ```suggestion
For example, Flask framework [promotes](https://flask-cors.readthedocs.io/en/latest/api.htm) origin reflection as the default behavior.
```

Reference won't work in hints, unfortunately.

{{< /hint >}}

If a resource hosted on `server.com` is requested from `target.com` then the origin could be reflected in the response headers as: `Access-Control-Allow-Origin: target.com`. If the resource is cached, this information is stored together with the resource in the browser cache. With that, if `attacker.com` tries to fetch the same resource there are two possible scenarios:
- The resource is not in cache: the resource could be fetched and stored together with the `Access-Control-Allow-Origin: attacker.com` header.
- The resource was already in cache: fetch attempt will try to fetch the resource from the cache but it will also generate a CORS error due to the ACAO header value mismatch with the requesting origin (`target.com` origin was expected but `attacker.com` was provided). Here below is provided an example code snippet epxloting this vulnerability to infer the cache status of the victim's browser. 
```javascript
// The function simply take a url and fetch it in CORS mode

> // if the fetch rise an error, it will be a CORS error due to the 

Reviewer: ```suggestion
// if the fetch raise an error, it will be a CORS error due to the 
```

// origin mismatch between attacker.com and victim's ip
function checkCachedResource(url) {
    fetch(url, {
        mode: "cors"
        }).catch((e) => {
            return true
        });
    return false
}

// This makes sense only if the attacker alredy knows that
// server.com suffers from origin reflection CORS misconfiguration
var resource_url = "server.com/reflected_origin_resource.html"
verdict = checkCachedResource(resource_url)
console.log("Resource was cached: " + verdict)
```

{{< hint tip >}}

> The best way to mitigate this is to avoid origin reflection and use the header `Access-Control-Allow-Origin: *` instead for globally accessible and unauthenticated resources.

Reviewer: ```suggestion
The best way to mitigate this is to avoid origin reflection and use the header `Access-Control-Allow-Origin: *` for globally accessible and unauthenticated resources.
```

[^2]: HTTP Cache Cross-Site Leaks, [link](http://sirdarckcat.blogspot.com/2019/03/http-cache-cross-site-leaks.html)
[^3]: Mass XS-Search using Cache Attack, [link](https://terjanq.github.io/Bug-Bounty/Google/cache-attack-06jd2d2mz2r0/index.html#VIII-YouTube-watching-history)
[^4]: Timing Attacks on Web Privacy, [link](http://www.cs.jhu.edu/~fabian/courses/CS600.424/course_papers/webtiming.pdf)
[^5]: CORS misconfiguration, [link](https://web-in-security.blogspot.com/2017/07/cors-misconfigurations-on-large-scale.html)

> [^6]: Python Flask `send_wildcard` default configuration, [link](https://flask-cors.readthedocs.io/en/latest/api.htm)

Reviewer: ```suggestion
```

See comments before.
